src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,175 @@
+  uiSettings: Record<string, any>;
+  plugins: PluginsServiceSetup;
+}
+
+interface PluginSpec {
+  readConfigValue: (config: any, key: any) => any;
+}
+
+interface Provider<T = object> {
+  fn: (
+    server: InternalHttpServiceSetup['server'],
+    config: ReturnType<PluginSpec['readConfigValue']>
+  ) => T;
+  pluginSpec: PluginSpec;
+}
+
+type App = ReturnType<InternalHttpServiceSetup['server']['getUiAppById']>;
+type Replacer<T = object> = (
+  variables: T,
+  request: KibanaRequest,
+  server: InternalHttpServiceSetup['server']
+) => Promise<T> | T;
+
+export class ApplicationContextProvider implements IApplicationContextProvider {
+  private readonly request!: KibanaRequest;
+  private readonly response!: KibanaResponseFactory;
+  private readonly env!: Env;
+  private readonly legacy!: LegacyServiceSetup;
+  private readonly http!: InternalHttpServiceSetup;
+  private readonly uiSettings!: Record<string, any>;
+  private readonly plugins!: PluginsServiceSetup;
+  private readonly template = compileFile(join(__dirname, 'views/ui_app.pug'));
+
+  constructor(params: Params) {
+    Object.assign(this, params);
+  }
+
+  private getCapabilities() {
+    const modifiers = this.http.server.getCapabilitiesModifiers();
+    const defaultCapabilties = this.http.server.getDefaultCapabilities();
+    const capabilities = mergeCapabilities(defaultCapabilties, {
+      // Get legacy nav links
+      navLinks: Object.assign(
+        {},
+        ...this.http.server.getUiNavLinks().map(({ _id }) => ({ [_id]: true }))
+      ),
+    });
+
+    return modifiers.reduce(
+      async (resolvedCapabilties, modifier) => modifier(this.request, await resolvedCapabilties),
+      Promise.resolve(capabilities)
+    );
+  }
+
+  private async getVariables(app: App, basePath: string, injectedOverrides: Record<string, any>) {
+    const providers: Provider[] = (this.legacy.uiExports as any).defaultInjectedVarProviders || [];
+    const replacers: Replacer[] = (this.legacy.uiExports as any).injectedVarsReplacers || [];
+    const defaultInjectedVars = providers.reduce(
+      (defaults, { fn, pluginSpec }) =>
+        mergeVariables(
+          defaults,
+          fn(this.http.server, pluginSpec.readConfigValue(this.http.server.config, []))
+        ),
+      {}
+    );
+    const appInjectedVars = await this.http.server.getInjectedUiAppVars(app.getId());
+    const injectedVars = mergeVariables(defaultInjectedVars, appInjectedVars, injectedOverrides);
+    const vars = await replacers.reduce(
+      async (variables, replacer) => await replacer(variables, this.request, this.http.server),
+      Promise.resolve(injectedVars)
+    );
+
+    return {
+      strictCsp: this.http.csp.strict,
+      uiPublicUrl: `${basePath}/ui`,
+      bootstrapScriptUrl: `${basePath}/bundles/app/${app.getId()}/bootstrap.js`,
+      i18n: i18n.translate,
+      locale: i18n.getLocale(),
+      darkMode: this.uiSettings?.user?.['theme:darkMode']?.userValue ?? false,
```

Comment:
This an indicator that we might have to keep those uiSettings in the core. @joshdover we don't have an issue to migrate them to the NP. It's something that we can keep in the LP until after 8.0.0